src/parse.yy

```diff
@@ -2813,53 +2819,70 @@ lAddDeclaration(DeclSpecs *ds, Declarator *decl) {
 }
 
 static void
-lAddTemplateDeclaration(TemplateParms *templateParmList, DeclSpecs *ds, Declarator *decl) {
-    if (ds == nullptr || decl == nullptr)
-        // Error happened earlier during parsing
-        return;
-
-    decl->InitFromDeclSpecs(ds);
-    if (ds->storageClass == SC_TYPEDEF) {
-        // FIXME: source position
-        Error(decl->pos, "Illegal \"typedef\" provided with function template.");
-        return;
+lCheckTemplateDeclSpecs(DeclSpecs *ds, SourcePos pos, TemplateType type, const char* name) {
+    std::string templateTypeStr;
+    switch (type) {
+        case TemplateType::Template:
+            templateTypeStr = "function template";
+            break;
+        case TemplateType::Instantiation:
+            templateTypeStr = "template instantiation";
+            break;
+        case TemplateType::Specialization:
+            templateTypeStr = "template specialization";
+            break;
+        default:
+            templateTypeStr = "function template";
+            break;
     }
     if (ds->typeQualifiers & TYPEQUAL_TASK){
-        Error(decl->pos, "'task' not supported for templates.");
+        Error(pos, "'task' not supported for %s.", templateTypeStr.c_str());
         return;
     }
     if (ds->typeQualifiers & TYPEQUAL_EXPORT) {
-        Error(decl->pos, "'export' not supported for templates.");
+        Error(pos, "'export' not supported for %s.", templateTypeStr.c_str());
+        return;
+    }
+    if (ds->storageClass == SC_TYPEDEF) {
+        // FIXME: source position
```

Comment:
What's wrong with source position here? Elaborate or remove.

I see that's coming from existing code, but nevertheless.